### PR TITLE
Implement ResizeObserver infra, support content box calculations. 

### LINKED ISSUE #31006: Implement ResizeObserver
Needed for https://github.com/servo/servo/issues/30822

Spec: https://drafts.csswg.org/resize-observer/#resize-observer-interface

One would have to add a new WebIDL(see `components/script/dom/webidl`), and expose the interface on `Window`. 

And then implement it in a new `components/script/dom/resizeobserver.rs` file. The WebIDL interface methods will be generated automatically and can be imported as `use crate::dom::bindings::codegen::Bindings::<interface name>`

For the callback part, `components/script/dom/eventtarget.rs` is probably a good place to look for inspiration.

### PR DESCRIPTION
**Pull request by gterzian**
Implement the basic infra for the [ResizeObserver](https://drafts.csswg.org/resize-observer/) Web API, and support size calculations for content boxes. 

---

- [ ] `./mach build -d` does not report any errors
- [ ] `./mach test-tidy` does not report any errors
- [ ] These changes fix #31006 (GitHub issue number if applicable)

- [ ] There are tests for these changes OR
- [ ] These changes do not require tests because ___

### CODE REVIEW

**Comment by gterzian** [comment]
[Those two](https://github.com/servo/servo/pull/31108#issuecomment-1911589318) appear to be intermittent, how do I mark them as such again?

I have filed an issue for tests that are still failing: https://github.com/servo/servo/issues/31182

The base case appears to work, there is some intermittency which appears related to the delivery of rafs.

**Comment by gterzian** [comment]
@mrobinson @jdm @emilio  This one is ready for review. See https://github.com/servo/servo/issues/31182 for the currently failing tests(usually it's one sub-test failing in the test)

Are the [tests failure](https://github.com/servo/servo/actions/runs/7695647188/job/20968986306#step:5:16) related to the switch of the main branch? 

> python: can't open file '/home/runner/work/servo/servo/./python/servo/try_parser.py': [Errno 2] No such file or directory

File: components/script/dom/resizeobserver.rs
Comment (Loirooriol) [inline components/script/dom/resizeobserver.rs:255]: Device_pixel_content_box should snap to device pixels.
Also, `bounding_content_box_or_zero` is the wrong choice if there are multiple fragments, we don't want the bounding rect. As per the current spec it should just be the 1st fragment, with the idea of extending to an array with all fragments in the future.

Comment (gterzian) [reply components/script/dom/resizeobserver.rs:255]: Thanks! I've added your comment to https://github.com/servo/servo/issues/31182 because I think some of these might best be addressed as follow-ups, so that this PR can focus more on the overal mechanism and get the event-loop integration right. 

Would you happen to have links to the code where one could get the right layout queries or transform the resulting data into the right format?

**Comment by gterzian** [comment]
I'm looking a bit more into this raf issue that makes a sub test fail, I guess it would be nice to have the entire `event-loop.html` test pass. Currently failing is one where it expects to have received a raf but it hasn't....

Feel free to review in the meantime...

File: components/script/dom/resizeobserver.rs
Comment (gterzian) [reply components/script/dom/resizeobserver.rs:216]: Yeah nothing... I was thinking of making this just an option, because only the last one is ever read. What do you think?

File: components/script/dom/resizeobserver.rs
Comment (gterzian) [reply components/script/dom/resizeobserver.rs:143]: I'm using the equivalent that isn't a dom object, so that `ResizeObservation` doesn't have to be a dom object either(which would with our codegen require exposing it to JS).

**Comment by gterzian** [comment]
Regarding the `event-loop.html` test: it does not receive a raf when it expects to. The tests [acknowledges](https://github.com/servo/servo/blob/9b6c473695e14c1a37dd70325657519b901e7efc/tests/wpt/tests/resize-observer/eventloop.html#L91) this potential issue, and tries to work around it, but I guess in our case this is not enough. I tried moving around the raf code in the script-thread and the test would pass, but I don't think we should look too much into this now. [test0](https://github.com/servo/servo/blob/9b6c473695e14c1a37dd70325657519b901e7efc/tests/wpt/tests/resize-observer/eventloop.html#L45) and [test2](https://github.com/servo/servo/blob/9b6c473695e14c1a37dd70325657519b901e7efc/tests/wpt/tests/resize-observer/eventloop.html#L196) pass, with the exeption of the raf count subtest. test1 seems to test shadow dom things, and it times out.

File: components/script/dom/resizeobserver.rs
Comment (emilio) [reply components/script/dom/resizeobserver.rs:216]: An option is fine as long as we only report one size at a time. you really want to set this to the array we last reported, rather than keeping appending.

Comment (gterzian) [reply components/script/dom/resizeobserver.rs:216]: Ah ok, now I understand: it's the array for multiple fragments for a single reported size, where currently one fragment is used. I thought it was more of a stack of reported sizes... thanks!

**Comment by gterzian** [comment]
@Loirooriol @emilio Thanks to the both of your for your initial comments, which have been addressed. 

Could I get a comprehensive review please(also perhaps from @mrobinson @jdm if you have time)? 

I think the current mechanism works, with two seperate areas that needs to be addressed:
1. Various size calculations are off, as pointed out by @Loirooriol 
2. We fail tests relying on proper integration with the raf callbacks. 

Both points are addressed in https://github.com/servo/servo/issues/31182

Regarding https://github.com/servo/servo/issues/31242, on which 2 above depends: I think merging this initial work is an excellent way to get feedback while trying to fix the event-loop in a follow-up, see the [currently failing test](https://github.com/servo/servo/blob/e15262c5a03f78f480282900d6828be19774ced3/tests/wpt/tests/resize-observer/eventloop.html#L83).

File: components/script/dom/resizeobserver.rs
Comment (gterzian) [inline components/script/dom/resizeobserver.rs:252]: @mrobinson this is the entry point for the layout calcs.

File: components/script/dom/resizeobserver.rs
Comment (Loirooriol) [inline components/script/dom/resizeobserver.rs:78]: This is per spec, but has memory problems: https://bugzilla.mozilla.org/show_bug.cgi?id=1596992

I don't remember how Blink managed to avoid that, but I think you should either do the same or disobey the spec like Gecko and WebKit do: https://github.com/w3c/csswg-drafts/issues/4518

Comment (gterzian) [reply components/script/dom/resizeobserver.rs:78]: Thanks for pointing it out, there is a note over at https://github.com/servo/servo/pull/31108/files#diff-b6bdaf97b920500ab94dea9827cda04ff8136b0f23bd0c515d9195c6be31a820R439

and I'll add the info you mention to it...

**Comment by gterzian** [comment]
@mrobinson Ping for the review items we discussed...

**Comment by mrobinson** [comment]
> @mrobinson Ping for the review items we discussed...

I've just posted #31696 which should allow using the correct bounding rectangle for this API.

**Review by mrobinson — CHANGES_REQUESTED** [review]
I think this just needs to use the new API for getting the boundaries of the first box for an element.

File: components/script/dom/document.rs
Comment (gterzian) [reply components/script/dom/document.rs:2906]: Why? It seems unnecessary to me since the whole `dom` module is not public.

File: components/script/dom/resizeobserver.rs
Comment (gterzian) [reply components/script/dom/resizeobserver.rs]: Yep, cleaned-up these.

File: components/script/script_thread.rs
Comment (jdm) [inline components/script/script_thread.rs]: Can we use a while here instead?

File: tests/wpt/meta/resize-observer/observe.html.ini
Comment (jdm) [inline tests/wpt/meta/resize-observer/observe.html.ini:2]: Do we know what the error is here?

File: tests/wpt/meta/resize-observer/notify.html.ini
Comment (jdm) [inline tests/wpt/meta/resize-observer/notify.html.ini:2]: Do we know what the error is here?

File: tests/wpt/meta/resize-observer/observe.html.ini
Comment (gterzian) [reply tests/wpt/meta/resize-observer/observe.html.ini:2]: Yes, it is a missing API. See https://github.com/servo/servo/issues/31182

File: components/script/dom/document.rs
Comment (gterzian) [reply components/script/dom/document.rs:2960]: When we call the observer callback, new observers can be registered, which would could a double-borrow(will add a more detailed comment).

File: components/script/dom/resizeobserver.rs
Comment (gterzian) [reply components/script/dom/resizeobserver.rs:143]: Will do.

File: components/script/dom/resizeobserver.rs
Comment (gterzian) [reply components/script/dom/resizeobserver.rs:150]: I've taken another look at all uses of `Call__` in the codebase, and we ignore all return values.

File: components/script/dom/resizeobserver.rs
Comment (gterzian) [reply components/script/dom/resizeobserver.rs:203]: ok

File: components/script/dom/resizeobserver.rs
Comment (gterzian) [reply components/script/dom/resizeobserver.rs:222]: I guess not, will look into.

File: components/script/script_thread.rs
Comment (gterzian) [reply components/script/script_thread.rs]: Will do.

**Comment by gterzian** [comment]
@jdm @Loirooriol @emilio @mrobinson As mentioned above, I would like to ask for a final review and try to merge this one, so that this would give us the basic infra(now integrated with "update the rendering" in the event-loop), and that remaining layout calculations, as well as a few other points noted in https://github.com/servo/servo/issues/31182, can be left as follow-ups.

**Review by jdm — APPROVED** [review]
Looks great! Thanks!

**Comment by gterzian** [comment]
@mrobinson I think the PR needs an approval from you as well. I have addressed your suggestions at the time, but I haven't added the additional layout calculations because I think it's a good follow-up(see comment above).

File: components/script/dom/resizeobserver.rs
Comment (mrobinson) [reply components/script/dom/resizeobserver.rs:266]: It should be possible to use the new query API that's been exposed to implement this correct versus in a followup

Comment (gterzian) [reply components/script/dom/resizeobserver.rs]: > use the new query API

Do you mean `content_boxes`? If so, I have added a commit that uses that one.

File: components/script/dom/resizeobserver.rs
Comment (gterzian) [reply components/script/dom/resizeobserver.rs]: Thank you.

File: components/script/dom/document.rs
Comment (mrobinson) [inline components/script/dom/document.rs:467]: URLs need to be wrapped in rustdoc comments or it will cause a warning:

```suggestion
    /// <https://drafts.csswg.org/resize-observer/#lifetime>.
    /// But implementing it comes with known problems:
    /// - <https://bugzilla.mozilla.org/show_bug.cgi?id=1596992>
    /// - <https://github.com/w3c/csswg-drafts/issues/4518>
```

File: components/script/dom/node.rs
Comment (mrobinson) [inline components/script/dom/node.rs:2117]:
```suggestion
        ScriptThread::note_rendering_opportunity(window_from_node(parent).pipeline_id());
```

File: components/script/dom/node.rs
Comment (mrobinson) [inline components/script/dom/node.rs:2245]:
```suggestion
        ScriptThread::note_rendering_opportunity(window_from_node(parent).pipeline_id());
```

File: components/script/script_thread.rs
Comment (mrobinson) [inline components/script/script_thread.rs:1991]:
```suggestion
               // Document load is a rendering opportunity.
                ScriptThread::note_rendering_opportunity(document.window().pipeline_id());
```

File: components/script/dom/document.rs
Comment (mrobinson) [inline components/script/dom/document.rs:2979]: obs isn't really a well known abbreviation, so I would avoid it:

```suggestion
            .any(|observer| observer.has_skipped_resize_observations())
```

File: components/script/dom/resizeobserver.rs
Comment (mrobinson) [inline components/script/dom/resizeobserver.rs:164]:
```suggestion
            .any(|(observeration, _)| observation.state == ObservationState::Skipped)
```

File: components/script/dom/resizeobserver.rs
Comment (mrobinson) [inline components/script/dom/resizeobserver.rs:144]:
```suggestion
            // initialized with one reported size (zero).
```

File: components/script/dom/resizeobserver.rs
Comment (mrobinson) [inline components/script/dom/resizeobserver.rs:156]: Should this only happen when the size changes? I'm a bit unclear on the mechanism here because is_active() checks to see if the size has changed, but it seems this code both updates the size and then calls the resize observer callbacks. I don't see how the code avoids that if the size changes.

It also seems like `is_active()` is calculating the size but not caching it anywhere...meaning it updates here as well. It's not a cheap calculating right now, unfortunately, so that seems a little wasteful. Perhaps I misunderstand this code though...?

File: components/script/dom/resizeobserver.rs
Comment (mrobinson) [inline components/script/dom/resizeobserver.rs:190]:
```suggestion
        ScriptThread::note_rendering_opportunity( window_from_node(target).pipeline_id());
```

File: components/script/dom/resizeobserver.rs
Comment (mrobinson) [inline components/script/dom/resizeobserver.rs:252]: I think the follow is shorter and clearer:
```suggestion
        box_size.width().to_f64_px() != last_reported_size.inline_size() ||
            box_size.height().to_f64_px() != last_reported_size.block_size()
```

**Review by mrobinson — CHANGES_REQUESTED** [review]
Just a couple nits, but I don't see any big issues with this change. Hopefully we can land it in the next couple days. Sorry for the big delays!

File: components/script/dom/resizeobserver.rs
Comment (gterzian) [reply components/script/dom/resizeobserver.rs:155]: > Should this only happen when the size changes?

This only happens for active observations, so that means size has changed plus observation depth is right. The code here does not update the size.

> It's not a cheap calculating right now, unfortunately, so that seems a little wasteful.

I have added a caching mechanism, see the last commit.

**Review by mrobinson — APPROVED** [review]

**Comment by mrobinson** [comment]
@gterzian Looks good, though please squash your changes before landing or else the commit message will contain a lot of noise. Thanks you!

**Comment by gterzian** [comment]
@Loirooriol @jdm @mrobinson thank you for your reviews.

**Pull request merged**
